### ISSUE TITLE
Friendly error output

### ISSUE THREAD

**Issue by gaearon** [ampedandwired/html-webpack-plugin#418]
We are big fans of this plugin and use it in [Create React App](https://github.com/facebookincubator/create-react-app).

The only thing where it doesn’t shine, in my opinion, is its error output. This might look great to the plugin developer, or even an experienced Webpack user, but it is terrifying to a beginner:

<img width="1438" alt="screen shot 2016-09-03 at 13 37 07" src="https://cloud.githubusercontent.com/assets/810438/18224848/8835c538-71db-11e6-83a6-e785078bbf1c.png">

In Create React App, we really care about clean and actionable error output.
Let’s take a closer look.

```
Error in   Error: Child compilation failed:
```

Error in Error? Child compilation failed? Both phrases are meaningless to the end user.

```
  Module not found: ./src/favicon.ic  xo in /Users/dan/p/create-react-app/template:
```

Something’s wrong with formatting: I requested `./src/favicon.icxo` but message puts a space in the filename.

```
  Error: Cannot resolve 'file' or 'directory' ./src/favicon.icxo in /Users/dan/p  /create-react-app/template
```

This line completely duplicates the previous one and is not useful.

  ```
  - compiler.js:76 
    [create-react-app]/[html-webpack-plugin]/lib/compiler.js:76:16
  
  - Compiler.js:214 Compiler.<anonymous>
    [create-react-app]/[webpack]/lib/Compiler.js:214:10
  
  - Compiler.js:403 
    [create-react-app]/[webpack]/lib/Compiler.js:403:12
  
  - Tapable.js:67 Compiler.next
    [create-react-app]/[tapable]/lib/Tapable.js:67:11
  
  - CachePlugin.js:40 Compiler.<anonymous>
    [create-react-app]/[webpack]/lib/CachePlugin.js:40:4
  
  - Tapable.js:71 Compiler.applyPluginsAsync
    [create-react-app]/[tapable]/lib/Tapable.js:71:13
  
  - Compiler.js:400 Compiler.<anonymous>
    [create-react-app]/[webpack]/lib/Compiler.js:400:9
  
  - Compilation.js:577 Compilation.<anonymous>
    [create-react-app]/[webpack]/lib/Compilation.js:577:13
  
  - Tapable.js:60 Compilation.applyPluginsAsync
    [create-react-app]/[tapable]/lib/Tapable.js:60:69
  
  - Compilation.js:572 Compilation.<anonymous>
    [create-react-app]/[webpack]/lib/Compilation.js:572:10
  
  - Tapable.js:60 Compilation.applyPluginsAsync
    [create-react-app]/[tapable]/lib/Tapable.js:60:69
  
  - Compilation.js:567 Compilation.<anonymous>
    [create-react-app]/[webpack]/lib/Compilation.js:567:9
  
  - Tapable.js:60 Compilation.applyPluginsAsync
    [create-react-app]/[tapable]/lib/Tapable.js:60:69
  
  - Compilation.js:563 Compilation.<anonymous>
    [create-react-app]/[webpack]/lib/Compilation.js:563:8
  
  - Tapable.js:60 Compilation.applyPluginsAsync
    [create-react-app]/[tapable]/lib/Tapable.js:60:69
  
  - Compilation.js:525 Compilation.seal
    [create-react-app]/[webpack]/lib/Compilation.js:525:7
  
  - From previous event:
  
  - compiler.js:69 Object.compileTemplate
    [create-react-app]/[html-webpack-plugin]/lib/compiler.js:69:10
  
  - index.js:47 Compiler.<anonymous>
    [create-react-app]/[html-webpack-plugin]/index.js:47:40
  
  - Tapable.js:107 Compiler.applyPluginsParallel
    [create-react-app]/[tapable]/lib/Tapable.js:107:14
  
  - Compiler.js:394 Compiler.compile
    [create-react-app]/[webpack]/lib/Compiler.js:394:7
  
  - Compiler.js:46 Watching.<anonymous>
    [create-react-app]/[webpack]/lib/Compiler.js:46:17
  
  - Tapable.js:67 Compiler.next
    [create-react-app]/[tapable]/lib/Tapable.js:67:11
  
  - middleware.js:80 Compiler.invalidAsyncPlugin
    [create-react-app]/[webpack-dev-middleware]/middleware.js:80:3
  
  - Tapable.js:71 Compiler.applyPluginsAsync
    [create-react-app]/[tapable]/lib/Tapable.js:71:13
  
  - Compiler.js:44 Watching._go
    [create-react-app]/[webpack]/lib/Compiler.js:44:16
  
  - Compiler.js:108 Watching.invalidate
    [create-react-app]/[webpack]/lib/Compiler.js:108:8
  
  - Compiler.js:93 Watching.<anonymous>
    [create-react-app]/[webpack]/lib/Compiler.js:93:8
  
  - NodeWatchFileSystem.js:43 NodeWatchFileSystem.<anonymous>
    [create-react-app]/[webpack]/lib/node/NodeWatchFileSystem.js:43:3
  
  - watchpack.js:121 EventEmitter._onTimeout
    [create-react-app]/[watchpack]/lib/watchpack.js:121:7
  ```

The stack has no valuable information to the end user. You can enable it for developers or debugging the plugin by adding a `debug` or `verbose` option to the plugin, or something like this.

It would be great to see this addressed so users of this plugin don’t feel intimidated.
cc @andreypopp who’s done similar work for `css-loader` and `babel-loader`.

PS We really love this plugin!

**Comment by zperrault** [comment]
@ampedandwired, I'd be happy to take a go at this if it is something you'd like to happen. 

/cc gaearon

**Comment by jantimon** [comment]
hey I am absolutely open for better error messages.. In fact there are two errors which happen often:

1) the child compilation fails (e.g wrong loader)
2) the eval of the child compilation fails (front-end only loader or similar)

I already tried to get a readable stack trace especially if you provide a JavaScript template but maybe a fresh pair of eyes can push this in a new level

**Comment by zperrault** [comment]
@jantimon I'm going to give this a go this week.

**Comment by jantimon** [comment]
@zperrault what's the current progress on this? do you need help?

**Comment by zperrault** [comment]
@jantimon I was looking at this last night and ended up hitting a wall. I had looked at @andreypopp's contributions to babel loader and CSS loader but since those are loaders and not plugins they weren't as much help as I was hoping. 

It would be simple enough to remove stack traces and ensure proper formatting or messages but I was hoping to get code frame working for where errors are happening to provide the same amount of context as things like babel loader. 

Right now my plan is to read through the project more to better understand how everything is working as I've realized I won't be able to adapt the solutions used in the loaders like I had originally thought. 

Is there a public Slack team you're a member of or a preferred method of contact should I have any specific questions?
